### ISSUE TITLE
Can archive-push block database shutdown?

### ISSUE THREAD

**Issue by pellepelster** [pgbackrest/pgbackrest#2204]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 
2.45

2. PostgreSQL version: 
14/15

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each: 
Alpine 3.18

4. Did you install pgBackRest from source or from a package?
From system package

7. Describe the issue:
This issue is likely more a configuration question. I currently use pgbackrest as backup solution for PostgreSQL configured in a pretty standard way:

```
archive_command = 'pgbackrest --config /rds/config/pgbackrest.conf --log-level-console=info --log-path=/rds/log --stanza=my-db-instance1 archive-push  %p'
archive_mode = on

max_wal_senders = 3
wal_level = replica
```

During database startup though there is an extra step where some user provisioning happens for the database (user creation, password changes and so on). To prevent any outside access during this process the database is started without a public listener

```
pg_ctl -D "${PG_DATA_DIR}" start --options="-c listen_addresses=''"
```

Then user provisioning happens, and finally the database is shut down and started for real

```
pg_ctl -D "${PG_DATA_DIR}" stop
```

The issue is that sometimes the stop times out after 60 seconds because the database is busy pushing archives (the reason why there are so many archive pushes is unclear at the moment)

```
waiting for server to shut down....2023-10-11 17:15:25.131 GMT [51] LOG:  received fast shutdown request
2023-10-11 17:15:25.141 GMT [51] LOG:  aborting any active transactions
2023-10-11 17:15:25.142 GMT [51] LOG:  background worker "logical replication launcher" (PID 59) exited with exit code 1
2023-10-11 17:15:25.143 GMT [53] LOG:  shutting down
2023-10-11 17:15:25.312 P00   INFO: pushed WAL file '000000010000002600000053' to the archive
2023-10-11 17:15:25.312 P00   INFO: archive-push command end: completed successfully (437ms)
2023-10-11 17:15:25.349 P00   INFO: archive-push command begin 2.45: [pg_wal/000000010000002600000054] --config=/rds/config/pgbackrest.conf --exec-id=88-84df60cf --log-level-console=info --log-path=/rds/log --pg1-path=/storage/data/rds-postgresql/14 --repo2-path=/storage/backup --repo2-type=posix --stanza=rds-postgresql
2023-10-11 17:15:25.736 P00   INFO: pushed WAL file '000000010000002600000054' to the archive
2023-10-11 17:15:25.736 P00   INFO: archive-push command end: completed successfully (389ms)
2023-10-11 17:15:25.750 P00   INFO: archive-push command begin 2.45: [pg_wal/000000010000002600000055] --config=/rds/config/pgbackrest.conf --exec-id=89-ebfd1580 --log-level-console=info --log-path=/rds/log --pg1-path=/storage/data/rds-postgresql/14 --repo2-path=/storage/backup --repo2-type=posix --stanza=rds-postgresql
2023-10-11 17:15:26.132 P00   INFO: pushed WAL file '000000010000002600000055' to the archive
2023-10-11 17:15:26.133 P00   INFO: archive-push command end: completed successfully (386ms)
.2023-10-11 17:15:26.146 P00   INFO: archive-push command begin 2.45: [pg_wal/000000010000002600000056] --config=/rds/config/pgbackrest.conf --exec-id=90-a6517ab3 --log-level-console=info --log-path=/rds/log --pg1-path=/storage/data/rds-postgresql/14 --repo2-path=/storage/backup --repo2-type=posix --stanza=rds-postgresql
2023-10-11 17:15:26.455 P00   INFO: pushed WAL file '000000010000002600000056' to the archive
2023-10-11 17:15:26.455 P00   INFO: archive-push command end: completed successfully (312ms)
2023-10-11 17:15:26.468 P00   INFO: archive-push command begin 2.45: [pg_wal/000000010000002600000057] --config=/rds/config/pgbackrest.conf --exec-id=91-51389798 --log-level-console=info --log-path=/rds/log --pg1-path=/storage/data/rds-postgresql/14 --repo2-path=/storage/backup --repo2-type=posix --stanza=rds-postgresql

[...]

pg_ctl: server does not shut down
```

My questions are:

* does a running archive-push command block database shutdown?
* Is it safe to temporarily disable the archive_command and not risk any data loss?
   * https://www.postgresql.org/docs/current/continuous-archiving.html says 
   > "If you are archiving via shell and wish to temporarily stop archiving, one way to do it is to set archive_command to the empty string (''). This will cause WAL files to accumulate in pg_wal/ until a working archive_command is re-established."
* Would it also be safe set archive_mode=off for the maintenance start?
  * Will archiving be resumed afterwards without data loss?
* I am wondering if manually triggering a checkpoint before the stop command could be a solution to make sure the shutdown happens timely.
  * I know I can extend the shutdown timeout but I would like to have a predictable shutdown behaviour

**Comment by dwsteele** [comment]
> does a running archive-push command block database shutdown?

Yes, because Postgres wants to archive all un-archived segments on shutdown. Shutdown `immediate` circumvents this but then you going into crash recovery on restart which is not great. `This is recommended only in emergencies` according to the Postgres docs.

> Is it safe to temporarily disable the archive_command and not risk any data loss? 

If you don't have a backup of this cluster then the WAL is useless anyway, i.e. this is a new cluster that is being boot-strapped for the first time. If this an existing cluster that is being restarted then, yeah, you would lose the ability to do PITR from your prior backup, which could result in data loss if you needed to restore to another system.

> Would it also be safe set archive_mode=off for the maintenance start? 

Same as above.

> I am wondering if manually triggering a checkpoint before the stop command could be a solution to make sure the shutdown happens timely

This will not have any effect at all. WAL archiving is independent of checkpointing.

**Comment by pgstef** [comment]
> During database startup though there is an extra step where some user provisioning happens for the database (user creation, password changes and so on). To prevent any outside access during this process the database is started without a public listener
> 
> ```
> pg_ctl -D "${PG_DATA_DIR}" start --options="-c listen_addresses=''"
> ```

Also, why wouldn't you simply reject the network connections from the pg_hba.conf? That would avoid restarting the database since only a reload would be needed then...

Finding out why you have so many WAL segments might be a good idea too. Looking at the number of .ready files in the pg_wal/archive_status before stopping the cluster will also give you more info.

There are some ways to speed up the archiving using pgBackRest async archiving (and multi-processes) + changing compress-type (to e.g. zst). But that's not the core issue imho.

**Comment by pellepelster** [comment]
Thanks for your thorough feedback, that answers a lot of open questions.  

You are also right on `pg_hba.conf` but for historical reasons in this particular use case the database must only answer on TCP when everything is properly set up.

Also many thanks for your work in pgBackRest!
